service/udp.go

```diff
@@ -475,7 +485,7 @@ func timedCopy(clientAddr net.Addr, clientConn net.PacketConn, targetConn *natco
 			if err != nil {
 				return onet.NewConnectionError("ERR_PACK", "Failed to pack data to client", err)
 			}
-			proxyClientBytes, err = clientConn.WriteTo(buf, clientAddr)
+			proxyClientBytes, err = clientConn.WriteToFrom(buf, clientAddr, proxyIP)
```

Comment:
This is getting complicated. I shouldn't have to care about the clientAddr and proxyIP to read the shadowsocks piping logic. I just need to know where to write to.

Perhaps we can introduce a Client Writer, which holds the client address and proxy IP and exposes a `write(buf)` method. This way the piping logic can call `clientWriter.Write(buf)`.

There is an issue with `debugUDPAddr`, which needs the client address, so we would have to keep passing it.

However, perhaps we implement a `ShadowsocksWriter.ReadFrom(targetConn)` method, which implements the forwarding of one packet and can call an internal `writeToClient(clientConn)` method that encapsulates the `clientAddr` and `proxyIP`. This would remove the need for the awkward unnamed function we have in the loop. We could potentially implement the ControlMessage logic there too, to reduce layering (it will be easy to refactor that later if needed).

You will need to move the handling of the expiration error to outside the call.

`ShadowsocksWriter` can hold `pkt`, `clientConn`, `clientAddr`, `proxyIP`.

I believe the new layering would separate concerns and improve readability and maintenance.

Reply:
OK, I've introduced `boundWriter and `shadowsocksUDPWriter` types to separate these concerns more clearly.